### PR TITLE
OAuth: Make sub claim required for generic oauth behind feature toggle

### PR DESCRIPTION
**Pull request by kalleep**
**What is this feature?**
Replacement for https://github.com/grafana/grafana/pull/84745.

As a first step we decided to require sub claim behind a feature toggle so we have a way to opt and give customers time to fix their oauth integrations.

I also decided to perform this check in oauth client instead of the generic oauth connector so we would catch other clients where this could happen as well.

Part of https://github.com/grafana/identity-access-team/issues/603

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/featuremgmt/registry.go [hunk 1]
```diff
@@ -1181,6 +1181,14 @@ var (
 			FrontendOnly: true,
 			Owner:        grafanaObservabilityMetricsSquad,
 		},
+		{
+			Name:              "oauthRequireSubClaim",
+			Description:       "Requre that sub claims is present in oauth tokens.",
```
Comment (eleijonmarck) [inline pkg/services/featuremgmt/registry.go:1186]:
```suggestion
			Description:       "Require that sub claims is present in oauth tokens.",
```

Comment (eleijonmarck) [reply pkg/services/featuremgmt/registry.go:1186]: need feature toggle generation again

File: pkg/services/authn/clients/oauth.go [hunk 2]
```diff
@@ -148,10 +150,13 @@ func (c *OAuth) Authenticate(ctx context.Context, r *authn.Request) (*authn.Iden
...
 
-	// Implement in Grafana 11
-	// if userInfo.Id == "" {
-	// 	return nil, errors.New("idP did not return a user id")
-	// }
+	if userInfo.Id == "" {
+		if c.features.IsEnabledGlobally(featuremgmt.FlagOauthRequireSubClaim) {
+			return nil, errOAuthUserInfo.Errorf("missing required sub claims")
+		} else {
+			c.log.FromContext(ctx).Warn("Missing sub claim, oauth authentication without a sub claim is deprecated and will be rejected in future versions.")
```
Comment (eleijonmarck) [inline pkg/services/authn/clients/oauth.go:157]:
```suggestion
			c.log.FromContext(ctx).Warn("Missing sub claim, oauth authentication without a sub claim is deprecated and will be rejected in future versions. Enable `oauthRequireSubClaim` featureToggle to user the sub claim")
```

**Review by Jguer — APPROVED** [review]
all good after fixes

File: pkg/services/authn/clients/oauth.go (hunk 2 above)
Comment (kalleep) [reply pkg/services/authn/clients/oauth.go:157]: I think this is a bit much information in a log line so will skip it.

**Review by mgyongyosi — APPROVED** [review]

**Pull request merged**